polkadot/service/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+					_ => {
+						debug!("Error adding transaction to the pool: {:?}", e);
+						None
+					},
+				}
+			}
+		} else {
+			debug!("Error decoding transaction");
+			None
+		}
+	}
+}
+
+impl Service {
+	/// Creates and register protocol with the network service
+	pub fn new(config: Configuration) -> Result<Service, error::Error> {
+		// Create client
+		let executor = polkadot_executor::Executor::new();
+		let mut storage = Default::default();
+		let key: AccountId = Keyring::One.into();
+
+		let genesis_config = GenesisConfig {
+			validators: vec![key.clone()],
+			authorities: vec![key.clone()],
+			balances: vec![(Keyring::One.into(), 1u64 << 63), (Keyring::Two.into(), 1u64 << 63)].into_iter().collect(),
+			block_time: 5,			// 5 second block time.
+			session_length: 720,	// that's 1 hour per session.
+			sessions_per_era: 24,	// 24 hours per era.
+			bonding_duration: 90,	// 90 days per bond.
+			approval_ratio: 667,	// 66.7% approvals required for legislation.
+		};
+		let prepare_genesis = || {
+			storage = genesis_config.genesis_map();
+			let block = genesis::construct_genesis_block(&storage);
+			storage.extend(additional_storage_with_genesis(&block));
+			(primitives::block::Header::decode(&mut block.header.encode().as_ref()).expect("to_vec() always gives a valid serialisation; qed"), storage.into_iter().collect())
+		};
+
+		let _keystore = Keystore::open(config.keystore_path.into()).map_err(::error::ErrorKind::Keystore)?;
+		let client = Arc::new(client::new_in_mem(executor, prepare_genesis)?);
+		let best_header = client.header(&BlockId::Hash(client.info()?.chain.best_hash))?.expect("Best header always exists; qed");
+		info!("Starting Polkadot. Best block is #{}", best_header.number);
+		let transaction_pool = Arc::new(Mutex::new(TransactionPool::new(config.transaction_pool)));
+		let transaction_pool_adapter = Arc::new(TransactionPoolAdapter {
+			pool: transaction_pool.clone(),
+			client: client.clone(),
+		});
+		let network_params = network::Params {
+			config: network::ProtocolConfig {
+				roles: config.roles,
+			},
+			network_config: config.network,
+			chain: client.clone(),
+			transaction_pool: transaction_pool_adapter,
+		};
+		let network = network::Service::new(network_params)?;
+
+		// Spin consensus service if configured
+		let consensus_service = if config.roles & Role::VALIDATOR == Role::VALIDATOR {
+			Some(consensus::Service::new(client.clone(), network.clone(), transaction_pool, &best_header))
+		} else {
+			None
+		};
+
+		let thread_client = client.clone();
+		let thread_network = network.clone();
+		let thread = thread::spawn(move || {
+			thread_network.start_network();
+			let mut core = Core::new().expect("tokio::Core could not be created");
+			// Main loop
+			loop {
+				let events = thread_client.import_notification_stream().map(|notification| {
+					thread_network.on_block_imported(&notification.header);
+				});
+				if let Err(_) = core.run(events.into_future()) {
```

Comment:
same, i think that this is a misuse of `into_future()` and the loop isn't strictly required